http-mclient/src/Network/HTTP/MClient.hs

```diff
@@ -138,16 +140,57 @@ send' transactionDecoder decoder request = do
   sendRequest sendRequest' transactionDecoder decoder request
 
 sendRequest
-  :: SendRequest
+  :: forall env a . Env env
+  => SendRequest
   -> TransactionDecoder a
   -> ResponseDecoder a
   -> HTTP.Request
   -> MIO env (Result a)
 sendRequest sendRequest' transactionDecoder decoder request =
-  either transactionDecoder decoder
-    <$> Exception.tryJust selectException (liftIO $ sendRequest' request)
+  either transactionDecoder decoder <$> retriableSendRequest
   where
-    selectException :: HTTP.HttpException -> Maybe HTTP.HttpExceptionContent
-    selectException = \case
-      (HTTP.HttpExceptionRequest _request content) -> pure content
-      _other                                       -> empty
+    retriableSendRequest :: MIO env (Either HTTP.HttpExceptionContent (HTTP.Response LBS.ByteString))
+    retriableSendRequest
+      = Retry.retrying
+        limitedBackoff
+        isRetriable
+      $ const (Exception.tryJust selectException (liftIO $ sendRequest' request))
+      where
+        limitedBackoff :: Retry.RetryPolicy
+        limitedBackoff = Retry.exponentialBackoff 50_000 <> Retry.limitRetries (convertImpure retryLimit)
+
+        isRetriable
+          :: Retry.RetryStatus
+          -> Either HTTP.HttpExceptionContent (HTTP.Response LBS.ByteString)
+          -> MIO env Bool
+        isRetriable Retry.RetryStatus{..} = either process (const (pure False))
+          where
+            process :: HTTP.HttpExceptionContent -> MIO env Bool
+            process error
+              = Log.info
+                ( "Retrying failed request due to "
+                  <> showc error
+                  <> " attempt "
+                  <> showc rsIterNumber
+                  <> " of "
+                  <> showc retryLimit
+                )
+              $> isRequestNotSent error
+              where
+                isRequestNotSent = \case
+                  HTTP.ConnectionTimeout            -> True
```

Comment:
@epicallan Can be sure this one actually gets thrown BEFORE the HTTP request made it to the server possilby resulting in a side effect? I'm not to sure. Same for others.

We can ONLY retry exceptions that guarantee he HTTP request never made it to the server.

Reply:
@mbj I am also not super certain. Let me see how to get to 100% certainity

Reply:
@mbj  AFAICT `HTTP.ConnectionTimeout` is only called within the connection acquisition phase of an HTTP transaction. Which is separate from the phase where we send out the entire request. 
The connection phase requires a host address, port & optional `Proxy-Authorization` header for Proxy connections in order to make a connection. 

So, we may or may not reach a server during a `ConnectionTimeout`. The bottom line is we never send out the entire request, we just attempt a hand-shake. IMO this is grounds for retrying.